component/publisher-client/src/main/java/org/wso2/am/analytics/publisher/client/MoesifClient.java

```diff
@@ -0,0 +1,206 @@
+                // This is enough to happen only at the first attempt.
+                keyRetriever.removeMoesifKeyFromMap(orgId);
+            }
+            currentAttempt -= 1;
+            MoesifClientContextHolder.PUBLISH_ATTEMPTS.set(currentAttempt);
+            try {
+                Thread.sleep(MoesifMicroserviceConstants.TIME_TO_WAIT_PUBLISH);
+                publish(builder);
+            } catch (MetricReportingException e) {
+                log.error("Failing retry attempt at Moesif client", e);
+            } catch (InterruptedException e) {
+                log.error("Failing retry attempt at Moesif client", e);
+            }
+        } else if (currentAttempt == 0) {
+            log.error("Failed all retrying attempts. Event will be dropped");
+        }
+    }
+
+    /**
+     * publish method is responsible for checking the availability of relevant moesif key
+     * and initiating moesif client sdk.
+     */
+    public void publish(MetricEventBuilder builder) throws MetricReportingException {
+        Map<String, Object> event = builder.build();
+        ConcurrentHashMap<String, String> orgIDMoesifKeyMap = keyRetriever.getMoesifKeyMap();
+
+        String orgId = (String) event.get(Constants.ORGANIZATION_ID);
+        String moesifKey;
+        if (orgIDMoesifKeyMap.containsKey(orgId)) {
+            moesifKey = orgIDMoesifKeyMap.get(orgId);
+        } else {
+            // store events with orgID that misses moesif keys,
+            // in the internal map inside a queueMissed.
+            // call the microservice when the scheduled time reaches.
+            // put the elements in queue missed to eventQueue.
+            missedEventHandler.putMissedEvent(builder);
+            return;
+        }
+
+        // init moesif api client
+        MoesifAPIClient client = keyRetriever.getMoesifClient(moesifKey);
+        APIController api = client.getAPI();
+
+        APICallBack<HttpResponse> callBack = new APICallBack<HttpResponse>() {
+            public void onSuccess(HttpContext context, HttpResponse response) {
+                int statusCode = context.getResponse().getStatusCode();
+                if (statusCode == 200) {
+                    log.debug("Event successfully published.");
+                } else if (statusCode >= 400 && statusCode < 500) {
+                    log.error("Event publishing failed. Moesif returned " + statusCode);
+                } else {
+                    log.error("Event publishing failed.Retrying.");
+                    doRetry(orgId, builder);
+                }
+            }
+
+            public void onFailure(HttpContext context, Throwable error) {
+                int statusCode = context.getResponse().getStatusCode();
+
+                if (statusCode >= 400 && statusCode < 500) {
+                    log.error("Event publishing failed. Moesif returned:", statusCode);
+                } else if (error != null) {
```

Comment:
what is the difference between error != null scenario and else path. which kind of scenarios are captured in else if block.

Reply:
Here the throwable is a JsonProcessingException.
So thought there is no need to retry in such case.